torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -0,0 +1,329 @@
+                        if param.grad.dtype is torch.float16:
+                            # TODO: if this does cause Overflow error, how should users handle this?
+                            param.grad = param.grad.coalesce()
+                        to_unscale = param.grad._values()
+                    else:
+                        to_unscale = param.grad
+
+                    per_device_and_dtype_grads[to_unscale.device][to_unscale.dtype].append(to_unscale)
+
+            for device, per_dtype_grads in per_device_and_dtype_grads.items():
+                for grads in per_dtype_grads.values():
+                    if grads[0].device.type == "cpu":
+                        self._foreach_non_finite_check_and_unscale_cpu_(
+                            grads,
+                            per_device_found_inf.get(device),
+                            per_device_inv_scale.get(device),
+                        )
+                    else:
+                        torch._amp_foreach_non_finite_check_and_unscale_(  # type: ignore
+                            grads,
+                            per_device_found_inf.get(device),
+                            per_device_inv_scale.get(device),
+                        )
+        return per_device_found_inf._per_device_tensors
+
+    def unscale_(self, optimizer: SGD) -> None:
+        if not self._enabled:
+            return
+
+        self._check_scale_growth_tracker("unscale_")
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if optimizer_state["stage"] is OptState.UNSCALED:
+            raise RuntimeError("unscale_() has already been called on this optimizer since the last update().")
+        elif optimizer_state["stage"] is OptState.STEPPED:
+            raise RuntimeError("unscale_() is being called after step().")
+
+        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
+        assert self._scale is not None
+        inv_scale = self._scale.double().reciprocal().float()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device=self._scale.device)
+
+        optimizer_state["found_inf_per_device"] = self._unscale_grads_(optimizer, inv_scale, found_inf, True)
+        optimizer_state["stage"] = OptState.UNSCALED
+
+        # Synchronize the detected inf across the ranks
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+        last_handle = None
+
+        for v in optimizer_state["found_inf_per_device"].values():
+            if v.device.type == "cpu":
+                v_on_cuda = v.cuda()
+                last_handle = dist.all_reduce(v_on_cuda, async_op=True, group=self.process_group)
+                v_on_cuda.cpu()
+            else:
+                last_handle = dist.all_reduce(v, async_op=True, group=self.process_group)
+
+        # Make sure that the calls are done before moving out.
+        if last_handle:
+            last_handle.wait()
```

Comment:
hmm, i guess this is sort of using an assumption of NCCL IIUC, i.e. if we submit comm ops on the same stream on the same nccl pg (which is what we do here) then nccl takes care to run those sequentially currently I believe, even if they are not data dependent ops. But I guess to be more backend agnostic or robust, it might be worth to collect all the futures (shouldn't be too many as they are per device) and wait them all with `torch.futures.wait_all`

but this can be done as a follow up, I think currently this should be fine

Reply:
done.